internal/app/go-filecoin/storage_market_connector/common.go

```diff
@@ -0,0 +1,290 @@
+	receiptChan := make(chan *types.MessageReceipt)
+	errChan := make(chan error)
+
+	err := <-pubErrCh
+	if err != nil {
+		return nil, err
+	}
+
+	go func() {
+		err := c.waiter.Wait(ctx, mcid, func(b *block.Block, message *types.SignedMessage, r *types.MessageReceipt) error {
+			receiptChan <- r
+			return nil
+		})
+		if err != nil {
+			errChan <- err
+		}
+	}()
+
+	select {
+	case receipt := <-receiptChan:
+		if receipt.ExitCode != 0 {
+			return nil, xerrors.Errorf("non-zero exit code: %d", receipt.ExitCode)
+		}
+
+		return receipt, nil
+	case err := <-errChan:
+		return nil, err
+	case <-ctx.Done():
+		return nil, xerrors.New("context ended prematurely")
+	}
+}
+
+func (c *connectorCommon) addFunds(ctx context.Context, from address.Address, addr address.Address, amount tokenamount.TokenAmount) error {
+	params, err := abi.ToEncodedValues(addr)
+	if err != nil {
+		return err
+	}
+
+	fromAddr, err := fcaddr.NewFromBytes(from.Bytes())
+	if err != nil {
+		return err
+	}
+
+	mcid, cerr, err := c.outbox.Send(
+		ctx,
+		fromAddr,
+		fcaddr.StorageMarketAddress,
+		types.NewAttoFIL(amount.Int),
+		types.NewGasPrice(1),
+		types.NewGasUnits(300),
+		true,
+		fcsm.AddBalance,
+		params,
+	)
+	if err != nil {
+		return err
+	}
+
+	_, err = c.wait(ctx, mcid, cerr)
+
+	return err
+}
+
+// SignBytes uses the local wallet to sign the bytes with the given address
+func (c *connectorCommon) SignBytes(ctx context.Context, signer address.Address, b []byte) (*smtypes.Signature, error) {
+	var fcSigner fcaddr.Address
+	var err error
+
+	if fcSigner, err = fcaddr.NewFromBytes(signer.Bytes()); err != nil {
+		return nil, err
+	}
+
+	fcSig, err := c.wallet.SignBytes(b, fcSigner)
+	if err != nil {
+		return nil, err
+	}
+
+	var sigType string
+	if signer.Protocol() == address.BLS {
+		sigType = smtypes.KTBLS
+	} else {
+		sigType = smtypes.KTSecp256k1
+	}
+	return &smtypes.Signature{
+		Type: sigType,
+		Data: fcSig[:],
+	}, nil
+}
+
+func (c *connectorCommon) GetBalance(ctx context.Context, addr address.Address) (storagemarket.Balance, error) {
+	var smState spasm.StorageMarketActorState
+	err := c.chainStore.GetActorStateAt(ctx, c.chainStore.Head(), fcaddr.StorageMarketAddress, &smState)
+	if err != nil {
+		return storagemarket.Balance{}, err
+	}
+
+	// Dragons: Balance or similar should be an exported method on StorageMarketState. Do it ourselves for now.
+	available, ok := spautil.BalanceTable_GetEntry(smState.EscrowTable, addr)
+	if !ok {
+		available = spaabi.NewTokenAmount(0)
+	}
+
+	locked, ok := spautil.BalanceTable_GetEntry(smState.LockedReqTable, addr)
+	if !ok {
+		locked = spaabi.NewTokenAmount(0)
+	}
+
+	return storagemarket.Balance{
+		Available: tokenamount.FromInt(available.Int.Uint64()),
+		Locked:    tokenamount.FromInt(locked.Int.Uint64()),
+	}, nil
+}
+
+func (c *connectorCommon) GetMinerWorker(ctx context.Context, miner address.Address) (address.Address, error) {
```

Comment:
In this case, we're actually using the plumbing method to actually get the worker. This method exists a) because it's needed to satisfy the storage provider and client api interfaces, and b) because we currently need to do address translation.